Implement lastLineBaseline for flex containers.

sammygill opened the pull request:
#### 7957e0705174b2517ea83858a019b4ed32ba5cfc
<pre>
Implement lastLineBaseline for flex containers.
<a href="https://bugs.webkit.org/show_bug.cgi?id=246201">https://bugs.webkit.org/show_bug.cgi?id=246201</a>
rdar://100882674

Reviewed by Alan Bujtas.

CSS Flexible Box Layout Module Level 1 defines the last baseline for
flex containers so that they may participate in last baseline alignment.

The spec provides steps to take when the inline axis of the container is
the same as the main axis as well as when it matches the cross axis.

RenderFlexibleBox::getBaselineChild was created to refactor the code so
that the pieces that are used by both firstLineBaseline and
lastLineBaseline could be shared. The behavior changes slightly based
on the alignment being used, so we need to pass that in as well.

The OrderIterator, which is used to iterate over flex items in the order
specified by the &quot;order&quot; property (if it is set), also needed to be
slightly modified so that it could be used in last baseline alignment.
This is because for last baseline alignment we need to first look at
the endmost item on the last line and work our way towards the beginning
of the line. To simplify this, OrderIterator was modified to provide a
reverse method that iterates both the OrderValues from the CSS Property
(so that the item in order last goes first) and also starts iterating
from the last child of the flex container.

Spec reference: <a href="https://www.w3.org/TR/css-flexbox-1/#flex-baselines">https://www.w3.org/TR/css-flexbox-1/#flex-baselines</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-flex-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-flex-004-expected.txt:
* Source/WebCore/rendering/OrderIterator.cpp:
(WebCore::OrderIterator::next):
(WebCore::OrderIterator::reverse):
* Source/WebCore/rendering/OrderIterator.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::firstLineBaseline const):
(WebCore::RenderFlexibleBox::lastLineBaseline const):
(WebCore::RenderFlexibleBox::getBaselineChild const):
(WebCore::RenderFlexibleBox::layoutBlock):
(WebCore::RenderFlexibleBox::layoutAndPlaceChildren):
* Source/WebCore/rendering/RenderFlexibleBox.h:
m_numberOfInFlowChildrenOnFirstLine and
m_numberOfInFlowChildrenOnLastLine have both been changed to
std::optional&lt;size_t&gt;. Instead of being -1 before the items have been
placed on the line, they will now be empty optionals. However, the
overall behavior remains the same and all the references to the
variables have been change to reflect the new type.

* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-table-001-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-table-002-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-table-003-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-001-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-002-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-003-expected.txt: Removed.
These are no longer needed since the behavior matches what is expected
in the root TestExpectations now.

Canonical link: <a href="https://commits.webkit.org/255575@main">https://commits.webkit.org/255575@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b82df9b83e59a1c68ab4da714423fee91162238

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92924 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2131 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23504 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102644 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96926 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2136 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30467 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85316 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98773 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98590 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1462 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79404 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28379 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83375 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71498 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36867 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34671 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18194 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3861 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38541 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40791 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40462 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37374 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.h:
```diff
@@ -231,7 +233,8 @@ class RenderFlexibleBox : public RenderBlock {
     mutable OrderIterator m_orderIterator { *this };
     int m_numberOfInFlowChildrenOnFirstLine { -1 };
-    
+    int m_numberOfInFlowChildrenOnLastLine { -1 };
```
magic number, pls use optional instead (std::optional<size_t>)

sammygill replied:
Changed both m_numberOfInFlowChildrenOnFirstLine and m_numberOfInFlowChildrenOnLastLine to std::optional<size_t>. I updated all of the reference also. Let me know if you see any issues with them.

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -282,15 +292,41 @@ std::optional<LayoutUnit> RenderFlexibleBox::firstLineBaseline() const
+    ASSERT(alignment == ItemPosition::Baseline || alignment == ItemPosition::LastBaseline);
+
+    RenderBox* baselineChild = nullptr;
+    OrderIterator itr = (alignment == ItemPosition::Baseline) ? m_orderIterator : m_orderIterator.reverse();
```
auto childIterator =

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -282,15 +292,41 @@ std::optional<LayoutUnit> RenderFlexibleBox::firstLineBaseline() const
+    RenderBox* baselineChild = nullptr;
+    OrderIterator itr = (alignment == ItemPosition::Baseline) ? m_orderIterator : m_orderIterator.reverse();
+
+    int childNumber = 0;
```
size_t?

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -282,15 +292,41 @@ std::optional<LayoutUnit> RenderFlexibleBox::firstLineBaseline() const
+    OrderIterator itr = (alignment == ItemPosition::Baseline) ? m_orderIterator : m_orderIterator.reverse();
+
+    int childNumber = 0;
+    for (RenderBox* child = itr.first(); child; child = itr.next()) {
```
auto*

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -282,15 +292,41 @@ std::optional<LayoutUnit> RenderFlexibleBox::firstLineBaseline() const
         return LayoutUnit { (mainAxisExtentForChild(*baselineChild) + baselineChild->logicalTop()).toInt() };
 
-    std::optional<LayoutUnit> baseline = baselineChild->firstLineBaseline();
+    std::optional<LayoutUnit> baseline = baselineChild->lastLineBaseline();
```
auto

sammygill replied:
Changed

sammygill committed:
Implement lastLineBaseline for flex containers.
https://bugs.webkit.org/show_bug.cgi?id=246201
rdar://100882674

Reviewed by Alan Bujtas.

CSS Flexible Box Layout Module Level 1 defines the last baseline for
flex containers so that they may participate in last baseline alignment.

The spec provides steps to take when the inline axis of the container is
the same as the main axis as well as when it matches the cross axis.

RenderFlexibleBox::getBaselineChild was created to refactor the code so
that the pieces that are used by both firstLineBaseline and
lastLineBaseline could be shared. The behavior changes slightly based
on the alignment being used, so we need to pass that in as well.

The OrderIterator, which is used to iterate over flex items in the order
specified by the "order" property (if it is set), also needed to be
slightly modified so that it could be used in last baseline alignment.
This is because for last baseline alignment we need to first look at
the endmost item on the last line and work our way towards the beginning
of the line. To simplify this, OrderIterator was modified to provide a
reverse method that iterates both the OrderValues from the CSS Property
(so that the item in order last goes first) and also starts iterating
from the last child of the flex container.

Spec reference: https://www.w3.org/TR/css-flexbox-1/#flex-baselines

* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-flex-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-flex-004-expected.txt:
* Source/WebCore/rendering/OrderIterator.cpp:
(WebCore::OrderIterator::next):
(WebCore::OrderIterator::reverse):
* Source/WebCore/rendering/OrderIterator.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::firstLineBaseline const):
(WebCore::RenderFlexibleBox::lastLineBaseline const):
(WebCore::RenderFlexibleBox::getBaselineChild const):
(WebCore::RenderFlexibleBox::layoutBlock):
(WebCore::RenderFlexibleBox::layoutAndPlaceChildren):
* Source/WebCore/rendering/RenderFlexibleBox.h:
m_numberOfInFlowChildrenOnFirstLine and
m_numberOfInFlowChildrenOnLastLine have both been changed to
std::optional<size_t>. Instead of being -1 before the items have been
placed on the line, they will now be empty optionals. However, the
overall behavior remains the same and all the references to the
variables have been change to reflect the new type.

* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-table-001-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-table-002-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-table-003-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-001-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-002-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-003-expected.txt: Removed.
These are no longer needed since the behavior matches what is expected
in the root TestExpectations now.

Canonical link: https://commits.webkit.org/255575@main
